### DISCUSSION TITLE
Continuing discussion on FsLab and F# Data Science tech

### DISCUSSION THREAD

**Discussion by dsyme** [fslaborg/FsLab#3] [General]
This is to continue the discussion with @muehlhaus, @siavash-babaei, @panesofglass, @matthewcrews, @Shmew, @kMutagene, @cartermp, @nhirschey, @bvenn, @WalternativE and others at https://github.com/fslaborg/zzarchive-FsLab/issues/137

Welcome all!

**Comment by dsyme** [comment]
The new website at https://fslab.org is just currently a generic display of the projects in the incubation space. @muehlhaus feel free to populate with alternative material as things come together using whatever web tech you like :-)

The current projects in this incubation space are

* https://github.com/fslaborg/XPlot
* https://github.com/fslaborg/Deedle
* https://github.com/fslaborg/FSharp.Charting
* https://github.com/fslaborg/RProvider

Note FSharp.Charting is windows-only and should be renamed and moved to fsprojects - we don't want any windows-only stuff in FsLab going forward.

XPlot will probably soon be better replaced by Plotly.NET

So this is really Quite An Empty Space. Let's fill it with goodness and make sure to also embrace other assets like FSharp.Data, Giraffe etc. which live outside this space.

**Comment by siavash-babaei** [comment]
Thank you @dsyme for quickly clearing the way and taking care of old stuff. 
@muehlhaus thank you - and your team - for taking on the leading role. When you get around to it, I am sure quite a few people would be interested in your ideas and plans, etc. even the jist of your brainstorming sessions ...
On a different note, I think one of the first things we need is: simple guide-lines and how-to documents for contributions at all levels both for FsLab and SciSharp. I think it would help people with all levels of expertise in various fields to make whatever small or large contributions they can make. With a smaller community, squeezing every resource as much as possible, would be even more vital.

**Comment by Shmew** [comment]
> Let's fill it with goodness

I think [flips](https://github.com/matthewcrews/flips) would be an excellent inclusion.

**Reply by WalternativE** [reply]
We can ask @matthewcrews if he'd fancy it being a part of FsLab 😊

**Reply by matthewcrews** [reply]
I'd be happy to offer it. We are currently working on a significant update to 3.0 which should set it up for extensibility going forward. I'll talk to my collaborators but I don't see that being an issue.

I would love for it to find a home here.

**Reply by muehlhaus** [reply]
Hi @matthewcrews, this would be fantastic... Thank you.

**Reply by dsyme** [reply]
@matthewcrews I invited you to be co-owner on this org, please by all means use the power :-)

**Comment by jwood803** [comment]
Awesome progress! I'll be glad to help contribute tutorials, samples, or anything else that may be needed to help get folks started.

**Reply by dsyme** [reply]
> @jwood803 Awesome progress! I'll be glad to help contribute tutorials, samples, or anything else that may be needed to help get folks started.

I'd like to start adding more F# samples to the SciSharp stack of tech (Tensorflow.NET is a good one to start with).  This will mostly be cloning C# samples. Would you like to work on that too? I've no particular timeline for this, it's just something I plan to do.

**Reply by zyzhu** [reply]
Once @WalternativE's Deedle notebook formatting extension https://github.com/WalternativE/Deedle.DotNet.Interactive.Extension is added to official repo, I plan to convert some existing docs into notebooks and add binder on the repo for new users to play with. More examples and use cases are welcome!

**Reply by jwood803** [reply]
@dsyme That's perfect! I've already started working on video material for Tensorflow.NET and Keras.NET so I can work the F# examples in, as well.

**Reply by kMutagene** [reply]
@zyzhu Hard agree on notebooks containing samples and tutorials, that is something we should contain in as many repos as possible. I would however suggest we still keep a static documentation besides that.

**Reply by WalternativE** [reply]
@zyzhu It's on my list of things I want to make better before I can add a PR. I gathered a bit more experience with building stable-ish extensions for .NET Interactive and with a bit of luck I can get hold of the MSFT Dataframe people (regarding some ideas for displaying tabular data in notebooks - there is a bit of code in their extensions I'd really like to talk about with them).

**Comment by dsyme** [comment]
I'd encourage everyone involved in FsLab to watch the SciSharp discussions, and one or more of these repos

* https://github.com/SciSharp/TensorFlow.NET/
* https://github.com/SciSharp/Keras.NET/
* https://github.com/SciSharp/Numpy.NET/

These are implemented in C# but we want them to have good F#/C# docs and experience, and we should still get involved since they do form an important entry point for Python-aware users into the F#/.NET data science ecosystem.

We want SciSharp and FsLab to work well together and that means good communication, cooperation and co-participation so lpease do watch those repos if you're interested in them

**Reply by dsyme** [reply]
I'd like to emphasise just how important I think the "python-like" experience is for F# data science and what a fantastic job SciSharp have done with this (cc @oceania2018).

People differ about whether they personally want `AllThePascalNaming` or `thing.max` Python naming but I can assure you there is a big difference in reaction from existing data scientists when the see the python names they know and love - and there is far less need to write documentation.  There are so many wins to this and is particularly nice for F# where `lowerCaseNames` are already common.

Basically, _if your F# data science library mimics a Python experience I strongly encourage you to consider Python naming for the library_.  Like all of TensorFlow.NET, Keras.NET, Numpy.NET and also DiffSharp. I'll likely be converting TorchSharp to this naming too.

The advantages of .NET are not in the naming, they are in the strong typing, excellent runtime, excellent interop and enterprise-depolyability.

Note that TensorFlow.NET uses `lowerCaseNames` and is used by Microsoft as the backing implementation for Microsoft.ML.Tensorflow.  So it's a real thing.

**Reply by cartermp** [reply]
I honestly think that F#-style naming is "attractive" enough such that it's not terrible to diverge a little bit from Python idioms so long as the overall name is still the same (e.g., `ndarray` instead of `floopyArray`).

**Reply by dsyme** [reply]
Yes, that's what we do in DiffSharp - close to Python - enough for knowledge to be relevant - but not slavishly so

**Reply by Oceania2018** [reply]
Just want to clarify: `Keras.NET` and `Numpy.NET` rely on appropriate python library installed. They're a light wrapper based on `python.net` interoper. The `TensorFlow.NET` with `TensorFlow.Keras` doesn't rely on any python environment, they're a deep wrapper on `tensorflow` standard `c` export library.

**Reply by WalternativE** [reply]
Thanks for the clarification!

**Reply by nhirschey** [reply]
The F# project here is a good place to contribute https://github.com/SciSharp/SciSharp-Stack-Examples/tree/master/src/TensorFlowNET.Examples.FSharp . In partiular, `Program.fs` needs to be updated to match the C# equivalent so that we can demo all the different examples once they are converted (C# Program.cs here: https://github.com/SciSharp/SciSharp-Stack-Examples/blob/master/src/TensorFlowNET.Examples/Program.cs )

**Reply by Oceania2018** [reply]
@nhirschey I think it would be nice if we provide two [implementation](https://github.com/SciSharp/SciSharp-Stack-Examples) in C# and F#.
![image](https://user-images.githubusercontent.com/1705364/101935034-0c2ec880-3ba4-11eb-89c7-4da08c05e42d.png)

**Reply by nhirschey** [reply]
@Oceania2018 yes, I like that idea of "C#, F#" side by side on the landing page. Very clear and succinct. 

I found that **Linear Regression in Eager mode** was relatively easy to port ([C#](https://github.com/SciSharp/SciSharp-Stack-Examples/blob/master/src/TensorFlowNET.Examples/BasicModels/LinearRegressionEager.cs),[F#](https://github.com/SciSharp/TensorFlow.NET#f-example)]. That is a testament to your good design.  I think the other models will be relatively straightforward as well.

Porting the model runner (Program.cs, etc.) looks a bit more involved. That could benefit from somebody here with good C# and F# skills (I do not know C#).

**Reply by Oceania2018** [reply]
@nhirschey Correct, we might have another F# version `Example Runner`. I'm poor at F#.

Fortunately, there is already a startup project for F#:
![image](https://user-images.githubusercontent.com/1705364/101946210-5029c980-3bb4-11eb-802c-25f55495028c.png)

**Reply by nhirschey** [reply]
We might want to consider what the best format for F# SciSharp STACK examples would be. If we mimic the current C# examples, then they will be as *.fs files in a compiled project. However, when I look at some of the sources for the C# examples (e.g., [bidirectional RNN](https://github.com/aymericdamien/TensorFlow-Examples/blob/master/tensorflow_v2/notebooks/3_NeuralNetworks/bidirectional_rnn.ipynb)), I start thinking that perhaps *.fsx scripts or .NET Interactive notebooks would be best. That could make it easier for python users to embrace the .NET technology.

**Reply by jwood803** [reply]
I think using .NET Interactive Notebooks would be the best bet for these. Would it help if there were both notebooks and a solution with project files? Or would that just be overkill?

**Reply by Oceania2018** [reply]
@nhirschey and @jwood803 are you interested in adding some content for F# in our the [docs protal](https://scisharp.github.io/tensorflow-net-docs/#/)?

**Reply by dsyme** [reply]
Cool, good idea.  Github source here: https://github.com/SciSharp/tensorflow-net-docs

**Comment by siavash-babaei** [comment]
![image](https://user-images.githubusercontent.com/40157200/101852649-2a58e200-3b2c-11eb-814f-35f1ce4c4796.png)

So, I took this of Hadley Wickham book: it pretty much specifies a normal data science workflow. For Import, Tidy, Transform, Visualize, and Communication steps, we got some tools: TypeProviders, Deedle, Plotly.NET, FSharp.Stats, etc. 

The sore thumb would be the Model step. More specifically, simplifying and making prototyping easy with quick turn around.

**Comment by kMutagene** [comment]
@dsyme To get people started adding documentation and tutorials, I think for these kinds of F#-first data science projects it is really important to get an easy way to include plots in the docs. We mentioned this first in https://github.com/fsprojects/FSharp.Formatting/issues/611 (cc @HLWeil), and we have a workaround that postprocesses the results to generate inline html, but that won't work with the awesome new `dotnet fsdocs watch`, which I think would be very important for first-time contributors especially. So how can we help on that end to get charting into FSharp.Formatting?

**Comment by dsyme** [comment]
> @dsyme To get people started adding documentation and tutorials, I think for these kinds of F#-first data science projects it is really important to get an easy way to include plots in the docs. We mentioned this first in fsprojects/FSharp.Formatting#611 (cc @HLWeil), and we have a workaround that postprocesses the results to generate inline html, but that won't work with the awesome new dotnet fsdocs watch, which I think would be very important for first-time contributors especially. So how can we help on that end to get charting into FSharp.Formatting?

I'll review the issue - thanks for the context

**Comment by dsyme** [comment]
> @zyzhu Hard agree on notebooks containing samples and tutorials, that is something we should contain in as many repos as possible. I would however suggest we still keep a static documentation besides that.

@kMutagene I've been considering accepting .ipynb as input to the `fsdocs` tool, so HTML, markdown and scripts can be generated from them, including checking and execution of F# code.

**Reply by kMutagene** [reply]
> I've been considering accepting .ipynb as input to the fsdocs tool, so HTML, markdown and scripts can be generated from them, including checking and execution of F# code.

that would be absolutely fantastic for that purpose!

**Comment by muehlhaus** [comment]
@siavash-babaei  That is a great idea. We would start creating a landing page. My suggestion would be to follow your (or Hadley Wickham) schema for a normal data science workflow:
1.	Our Mission: FsLabs: Because Data Science hearts F# first (here we should come up with a nice slogan and some good explanation of the incubation space) 
2.	We would suggest F# (first) tools for every step in the schema predominantly from the FsLabs Stack
3.	Getting started with F# data science (tutorial + learning material)
4.	Make the connection to SciSharp
For the learning material I would like to come up with something focusing on people starting in F# as their first programming experience. This might take some time to prepare, but I think this is important. I not aware if something like that is freely available. Maybe I just missed it.
For experienced programmer I would point to some existing excellent material.
This would be my suggestion. But of course, this open for input and discussion

**Reply by dsyme** [reply]
> predominantly from the FsLabs Stack

Suggest a policy more of "we recommend the best tool for use from F# wherever it comes from" rather than prioritizing FsLabs-hosted projects.  Many (if not most) components will come from outside this incubation space, and you'll make friends with other communities/projects this way, and give yourself maximum flexibility for the future.

@muehlhaus I have the twitter details - will send them to you

**Reply by muehlhaus** [reply]
I completely agree... I just meant in case of a redundancy in functionality; we might prefer the one from the stack… but you are absolutely right and we will state it accordingly…

Thanks for the twitter details

**Reply by muehlhaus** [reply]
We started to built the landing page using the [fslaborg/fslabsite] repro. We will add content within the next days and than move it to fslab.org (fslaborg.github.io).

**Reply by dsyme** [reply]
Yup I saw the repo transfer in - this is great news!

**Reply by siavash-babaei** [reply]
Incidently, that graph of a typical data science workflow by Hadley Wickham somewhere up there is a very good blueprint. FsLab could organise tools according to that, e.g.,

1- XYZ-1 Type Provider for Data Access Step. FSharp.Data
2- XYZ-2 Package for Tidy Step. Deedle
3- XYZ-3 Package for Visualisation Step. Plotly
4- XYZ-4 Package for Exploration Step. FSharp.Stats
5- XYZ-5 Package for Modelling Step.
6- XYZ-6 Package for Communication Step. F# Notebooks
7- XYZ-7 Package for Cloud Computing.
...

My suggestion would be to prioritise to cover the areas that are most often used and where F# is already really good or decent at. Being able to neatly and quickly prototype a Regression model or a Decision tree or a Survival model is much more important than having facilities for more niche areas such as deep learning,  copula functions, or extreme value theory. As FSharp.Data and Deedle do indeed show-case: It can already be quite brilliant for accessing and wrangling/manipulating/tidying data.

There're good stuff already there needing update, consolidation, documentation. But I don't think we are that far off having a proper FsLab. Done properly, it will not only greatly help current user base but invite others to join in.

**Comment by dsyme** [comment]
I'd like to welcome @luisquintanilla to FsLab.  

Luis works on .NET Documentation for F# machine learning and much more at Microsoft and is passionate about F# in the data-science and machine learning space.

Luis is happy to help co-admin in FsLab

**Comment by luisquintanilla** [comment]
Thanks @dsyme 

Looking forward to working with the rest of the community on these projects 🙂

**Comment by smoothdeveloper** [comment]
I'd like the organization to help with consolidating the experience of the libraries in the different environments (.net interactive, fsi, etc.).

I'm taking the exemple of Flips which I'm getting acquainted with, it would be good to be able to print the decisions and equations into those environments in ways that make them more appealing to end users.

Those efforts can be consolidated if some of the people working at microsoft on the projects where this integrates / adds value, by contributing to individual projects with the proposed extensions.

There are also few other things that would be good for making fslab more attractive:

* define a format for showcasing the libraries, so fslaborg website would become an entry point to explore the A,B,Cs of each library
* consolidating / tailoring the format of the documentation we get from fsdocs, increasing the bar of the documentation with code reviews
* engage with fsharp.org so that mentee / mentors / interns can pick from a set of "approved by fslab" tasks and involve contribution from larger community

Those things would need to function on their own, not consume too much of the libraries maintainers time, wether or not the library is incubating or has a separate org.

**Reply by dsyme** [reply]
I'd really like to see FsLab contributors active in taking FSharp.Formatting/fsdocs forward, and also .NET Interactive notebooks too (including the support in Visual Studio code)

The fsdocs F# literate scripting programming model is good for basic script-based literate programming but it can really be improved through a few more iterations.  For example taking notebooks as input (described above), and also `fsdocs build --watch` should be usable as a pracitcal literate programming "watch my scripts and process them as I develop" tool.

**Reply by smoothdeveloper** [reply]
Adding to some of the things mentioned above:

## Library design guidelines

Similar to the work done on diffsharp or some of the Chart API which are assume are large and evolving with new options added in their backends.

Those guidelines would come handy for library designers that fit the domain space, and they'd be more specific and maybe different than the more general F# component design guidelines.

## End to end example project

Either made off a real world project that can be boiled down to more basic architecture and scale suitable for introduction material and additional contents (series of articles).

Those would be possibly be prioritized:
* code clarity
* usage of the libraries
* the clarity of the algorithms
* "look and feel" of the program to draw attention

## Help people coming from other environments

People with experience in the field could contribute a list of existing libraries / domains in the other eco-system, and help to map with the space from julia/python/R -> F#.

The website could publish this in a nice format with additional information.

## Find and seed more projects

There are probably F# datascience related libraries in the wild that aren't under any organization, or looking to contribute to the fslab website, and based on other points, see if additional projects should be created to cover specific domain / usages.


@dsyme:

> I'd really like to see FsLab contributors active in taking FSharp.Formatting/fsdocs forward, and also .NET Interactive notebooks too (including the support in Visual Studio code)
> The fsdocs F# literate scripting programming model is good for basic script-based literate programming but it can really be improved through a few more iterations. For example taking notebooks as input (described above), and also fsdocs build --watch should be usable as a pracitcal literate programming "watch my scripts and process them as I develop" tool.

Maybe creating a top level issue upstream about the extensibility requirements (support for notebook formats and others), fsdocs is potentially a place where lots of things could indeed happen for several aspects of what's under fslab (website, documentation of libs, and suitable to some of the datascientist usecases around pretty printing stuff).

I'm not familiar with the fsharp.formatting codebase, we may look if anything datascience related is done in the other ecosystems, for example in [pandoc](https://pandoc.org/) which supports jupyter notebook for both input and output.

**Reply by dsyme** [reply]
Yes, pandoc is a good reference point

**Reply by smoothdeveloper** [reply]
On slide 18 at https://youtu.be/PEoPrMkg8W4?t=1363 there is a mention of rust doc tests which I think is relevant to considerations around fsdocs.

**Comment by siavash-babaei** [comment]
One important thing that is sorely lacking is proper Vidro Tutorials. Checking DataCamp or Udemy, one can find thousands of hours of courses for say python. They make adoption and uptake real easy. Furthermore, say when a general programmer already familiar with core language and concepts, wants to start doing data science, it would take him/her tens of hours rather than hundreds to start being really productive in the field. I reckon around 150 hours of course would cover most thing from basics to advanced, concurrency, parallelism, design, backend, frontend, type providers, computation expressions, data science, etc. 

What few we currently have, scattered over Udemy, Pluralsight, FsharpTV, etc., are commonly very short, old, often only covering the very basics, and sometimes incomplete. Producing material is neither cheap nor easy, but those who can, are already donating significant time and effort. This has just as much effect - if not more - than code contribution or whatever: DSyme's Expert F# or Scott Wlaschin's site is as an important asset as any library or framework out there and I would feel just as giddy with an F# Spark as if "DDD with F#" was turned into an applied course. 

A good FsLab I think should include neatly organized tutorials for the various tasks and workflows, hopefully in Video as well as textual formats.

**Reply by muehlhaus** [reply]
I like your video idea very much. If you have a collection fitting to FsLab please just contribute it either directly or let me know. We have setup PR templates to extend the tutorials and/or the rest of the webpage. 

Your contribution is highly appreciated.

**Comment by siavash-babaei** [comment]
I saw this somewhere and thought "Hoom, Interesting Point":

"The problem with the F# library ecosystem is that it is VERY insular. People write libraries in F# for F#. Writing in F# is great. However most libraries should be written for .Net instead: the traction would be much larger and .Net is a perfectly good API target. While F#-F# sometimes results in better APIs (e.g. Fabulous), the restriction to .Net APIs would improve most libraries by preventing excesses (a common culprit is crazy operators).

Libraries which are written for F# which should be written for .Net include: FAKE, Farmer, most type providers (the erasing ones). If these were retargeted it would 1. help them remain actively maintained projects 2. integrate F# by bringing in contributors who normally work in other languages and turning F# into a useful producer language not just a consumer one."
